### PR TITLE
test: improve fungible module coverage

### PR DESCRIPTION
**Pull request by skosito**
# Description

Increases the coverage of mentioned parts of fungible module. Adds some generic helper methods for mocking evm client. I think there is a bit more room to move some testing code around (for example deploySystemContract) and probably it's better to mock happy paths too. Here I basically had to introduce more evm client mocking because it is hard to mimic real world evmCall reverts (and no need to in unit tests, we can mock that external dependency).

Relates to: #1742 . It is part 1 of adding more coverage, will add missing tests for `system_contract.go` in follow up PR.

## Type of change

Tests

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [x] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/fungible/keeper/evm_test.go [hunk 1]
```diff
@@ -33,20 +31,30 @@ func getValidChainID(t *testing.T) int64 {
...
-func assertContractDeployment(t *testing.T, k *evmkeeper.Keeper, ctx sdk.Context, contractAddress common.Address) {
+func assertContractDeployment(t *testing.T, k types.EVMKeeper, ctx sdk.Context, contractAddress common.Address) {
 	acc := k.GetAccount(ctx, contractAddress)
 	require.NotNil(t, acc)
 
 	code := k.GetCode(ctx, common.BytesToHash(acc.CodeHash))
 	require.NotEmpty(t, code)
 }
 
+func deploySystemContractsWithMockEvmKeeper(
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/fungible/keeper/evm_test.go:42]: >als ideally in unit tests only version with mocked evm keeper should be used i believe

It's something we should explore for improvement, in some cases this makes the testing easier even though it becomes more like an integration test

File: x/fungible/keeper/gas_price.go [hunk 2]
```diff
@@ -28,9 +31,6 @@ func (k Keeper) SetGasPrice(ctx sdk.Context, chainid *big.Int, gasPrice *big.Int
 	if err != nil {
 		return 0, sdkerrors.Wrapf(types.ErrContractCall, err.Error())
 	}
-	if res.Failed() {
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/fungible/keeper/gas_price.go:31]: Can we create an issue for this and avoid doing this change in this method? I may think I had a situation where `err == nil && res.Failed()`

File: x/fungible/keeper/msg_server_update_system_contract_test.go [hunk 3]
```diff
@@ -74,6 +75,41 @@ func TestKeeper_UpdateSystemContract(t *testing.T) {
...
+		k, ctx, _, zk := keepertest.FungibleKeeper(t)
+		msgServer := keeper.NewMsgServerImpl(*k)
+		k.GetAuthKeeper().GetModuleAccount(ctx, types.ModuleName)
+		admin := sample.AccAddress()
+		setAdminPolicies(ctx, zk, admin, observertypes.Policy_Type_group2)
+
+		chains := zetacommon.DefaultChainsList()
+		require.True(t, len(chains) > 1)
+		require.NotNil(t, chains[0])
+		require.NotNil(t, chains[1])
```
Comment (lumtis) [inline x/fungible/keeper/msg_server_update_system_contract_test.go:88]: Seems not used, why are we checking this?

File: x/fungible/keeper/msg_server_update_system_contract_test.go [hunk 4]
```diff
@@ -74,6 +75,41 @@ func TestKeeper_UpdateSystemContract(t *testing.T) {
...
+
+		// can update the system contract
+		_, err = msgServer.UpdateSystemContract(ctx, types.NewMsgUpdateSystemContract(admin, newSystemContract.Hex()))
+		require.NoError(t, err)
+
+		// can retrieve the system contract
+		sc, found := k.GetSystemContract(ctx)
+		require.True(t, found)
+		require.Equal(t, newSystemContract.Hex(), sc.SystemContract)
+	})
```
Comment (lumtis) [inline x/fungible/keeper/msg_server_update_system_contract_test.go:111]: Maybe we can deploy another system contract and check the previous one can be overwritten

File: x/fungible/keeper/msg_server_update_system_contract_test.go [hunk 5]
```diff
@@ -109,4 +145,69 @@ func TestKeeper_UpdateSystemContract(t *testing.T) {
...
+		mockEVMKeeper := keepertest.GetFungibleEVMMock(t, k)
+		msgServer := keeper.NewMsgServerImpl(*k)
+		k.GetAuthKeeper().GetModuleAccount(ctx, types.ModuleName)
+		admin := sample.AccAddress()
+		setAdminPolicies(ctx, zk, admin, observertypes.Policy_Type_group2)
+
+		chains := zetacommon.DefaultChainsList()
+		require.True(t, len(chains) > 1)
+		require.NotNil(t, chains[0])
+		require.NotNil(t, chains[1])
```
Comment (lumtis) [inline x/fungible/keeper/msg_server_update_system_contract_test.go:164]: This one looks not used?

**Review by lumtis — COMMENTED** [review]
Looks great!

File: testutil/keeper/fungible.go [hunk 6]
```diff
@@ -164,8 +169,73 @@ func GetFungibleObserverMock(t testing.TB, keeper *keeper.Keeper) *fungiblemocks
...
+func GetFungibleEVMMock(t testing.TB, keeper *keeper.Keeper) *FungibleMockEVMKeeper {
 	fek, ok := keeper.GetEVMKeeper().(*fungiblemocks.FungibleEVMKeeper)
 	require.True(t, ok)
-	return fek
+	return &FungibleMockEVMKeeper{
+		FungibleEVMKeeper: fek,
+	}
+}
+
+type FungibleMockEVMKeeper struct {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply testutil/keeper/fungible.go:180]: Looks like a good idea ,

**Review by lumtis — APPROVED** [review]
Great job!
The coverage looks good to me. We're integrating code coverage so will check the global coverage afterward.

**Pull request merged**